ftp.go

```diff
@@ -238,17 +247,29 @@ func (c *ServerConn) openDataConn() (net.Conn, error) {
 // cmd is a helper function to execute a command and check for the expected FTP
 // return code
 func (c *ServerConn) cmd(expected int, format string, args ...interface{}) (int, string, error) {
-	_, err := c.conn.Cmd(format, args...)
+	id, err := c.conn.Cmd(format, args...)
 	if err != nil {
 		return 0, "", err
 	}
 
+	// Utilize the Pipeline on c.conn to be safe for concurrency.
+	c.conn.StartResponse(id)
+	defer c.conn.EndResponse(id)
+
 	return c.conn.ReadResponse(expected)
 }
 
 // cmdDataConnFrom executes a command which require a FTP data connection.
 // Issues a REST FTP command to specify the number of bytes to skip for the transfer.
+//
+// Caller MUST call c.conn.EndResponse(c.id) AND c.mu.Unlock() once the returned
+// connection is closed.
 func (c *ServerConn) cmdDataConnFrom(offset uint64, format string, args ...interface{}) (net.Conn, error) {
+	// Lock ServerConn so that the data action can run through its synchronous
+	// sequence of events. Must be Unlock()'d before any other data action can
+	// proceed.
+	c.mu.Lock()
+
```

Comment:
The corresponding `c.mu.Unlock()` is dependent on the calling method. For instance, the `StorFrom()` method reads the data it needs from the connection, closes the connection, and is done with that transaction, so to speak. So that method is okay to `defer c.mu.Unlock()`. However, in the case of `RetrFrom()`, where it returns a `response` struct as its `io.ReadCloser`, the `c.mu.Unlock()` happens when `Close()` is called on the `response` since that is closed by code outside of this package.

It comes down to needing to lock the entire transaction around the `net.Conn` being opened, any data being read or written to it, and it being closed. Otherwise, deadlock occurs.